// omg/lib/commands/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "build", {
  enumerable: true,
  get: function () {
    return _build.default;
  }
});
Object.defineProperty(exports, "list", {
  enumerable: true,
  get: function () {
    return _list.default;
  }
});
Object.defineProperty(exports, "run", {
  enumerable: true,
  get: function () {
    return _run.default;
  }
});
Object.defineProperty(exports, "subscribe", {
  enumerable: true,
  get: function () {
    return _subscribe.default;
  }
});
Object.defineProperty(exports, "ui", {
  enumerable: true,
  get: function () {
    return _ui.default;
  }
});
Object.defineProperty(exports, "validate", {
  enumerable: true,
  get: function () {
    return _validate.default;
  }
});

var _build = _interopRequireDefault(require("./build"));

var _list = _interopRequireDefault(require("./list"));

var _run = _interopRequireDefault(require("./run"));

var _subscribe = _interopRequireDefault(require("./subscribe"));

var _ui = _interopRequireDefault(require("./ui"));

var _validate = _interopRequireDefault(require("./validate"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
