// omg/lib/services/config/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getContainerPorts", {
  enumerable: true,
  get: function () {
    return _getContainerPorts.default;
  }
});
Object.defineProperty(exports, "getConfigPaths", {
  enumerable: true,
  get: function () {
    return _getConfigPaths.default;
  }
});
Object.defineProperty(exports, "ConfigPaths", {
  enumerable: true,
  get: function () {
    return _getConfigPaths.ConfigPaths;
  }
});
Object.defineProperty(exports, "getValidationErrors", {
  enumerable: true,
  get: function () {
    return _getValidationErrors.default;
  }
});
Object.defineProperty(exports, "parseMicroserviceConfig", {
  enumerable: true,
  get: function () {
    return _parseMicroserviceConfig.default;
  }
});
Object.defineProperty(exports, "watchConfigFile", {
  enumerable: true,
  get: function () {
    return _watchConfigFile.default;
  }
});

var _getContainerPorts = _interopRequireDefault(require("./getContainerPorts"));

var _getConfigPaths = _interopRequireWildcard(require("./getConfigPaths"));

var _getValidationErrors = _interopRequireDefault(require("./getValidationErrors"));

var _parseMicroserviceConfig = _interopRequireDefault(require("./parseMicroserviceConfig"));

var _watchConfigFile = _interopRequireDefault(require("./watchConfigFile"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
