// omg/lib/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.disposableInterval = disposableInterval;
exports.setCliOptions = setCliOptions;
exports.getCliOptions = getCliOptions;
exports.HELP_OMG_LIST_DETAILS = exports.HELP_OMG_LIST = exports.lifecycleDisposables = exports.DEBUG_CLI = void 0;

var _eventKit = require("event-kit");

const DEBUG_CLI = {}.hasOwnProperty.call(process.env, 'OMS_CLI_DEBUG');
exports.DEBUG_CLI = DEBUG_CLI;
const lifecycleDisposables = new _eventKit.CompositeDisposable();
exports.lifecycleDisposables = lifecycleDisposables;

function disposableInterval(callback, timeout) {
  const intervalId = setInterval(callback, timeout);
  return new _eventKit.Disposable(() => {
    clearInterval(intervalId);
  });
}

let options = {};

function setCliOptions(newOptions) {
  options = newOptions;
}

function getCliOptions() {
  return options;
}

const HELP_OMG_LIST = `Try 'omg list' to get a list of available actions`;
exports.HELP_OMG_LIST = HELP_OMG_LIST;
const HELP_OMG_LIST_DETAILS = `Try 'omg list --details' to get a detailed list of available actions`;
exports.HELP_OMG_LIST_DETAILS = HELP_OMG_LIST_DETAILS;
